src/outputs/parthenon_hdf5.cpp

```diff
@@ -15,17 +15,193 @@
 // the public, perform publicly and display publicly, and to permit others to do so.
 //========================================================================================
 
+// options for building
+#include "config.hpp"
+
 // Only proceed if HDF5 output enabled
+#ifdef HDF5OUTPUT
 
+#include <algorithm>
 #include <memory>
+#include <set>
+#include <type_traits>
+#include <unordered_map>
 
+#include "interface/meshblock_data_iterator.hpp"
+#include "interface/metadata.hpp"
+#include "mesh/mesh.hpp"
+#include "mesh/meshblock.hpp"
+#include "outputs/outputs.hpp"
 #include "outputs/parthenon_hdf5.hpp"
 
-#include "mesh/meshblock.hpp"
+namespace parthenon {
+namespace HDF5 {
+
+// template specializations for std::string
+template <>
+void HDF5WriteAttribute(const std::string &name, const std::vector<std::string> &values,
+                        hid_t location) {
+  std::vector<const char *> char_ptrs(values.size());
+  for (size_t i = 0; i < values.size(); ++i) {
+    char_ptrs[i] = values[i].c_str();
+  }
+  HDF5WriteAttribute(name, char_ptrs, location);
+}
 
-#ifdef HDF5OUTPUT
+template <>
+std::vector<std::string> HDF5ReadAttributeVec(hid_t location, const std::string &name) {
+  // get strings as char pointers, HDF5 will allocate the memory and we need to free it
+  auto char_ptrs = HDF5ReadAttributeVec<char *>(location, name);
 
-namespace parthenon {
+  // make strings out of char pointers, which copies the memory and then free the memeory
+  std::vector<std::string> res(char_ptrs.size());
+  for (size_t i = 0; i < res.size(); ++i) {
+    res[i] = std::string(char_ptrs[i]);
+    free(char_ptrs[i]);
+  }
+
+  return res;
+}
+
+// template specialization for bool
+template <>
+void HDF5WriteAttribute(const std::string &name, const std::vector<bool> &values,
+                        hid_t location) {
+  // can't use std::vector here because std::vector<bool>  doesn't have .data() member
+  std::unique_ptr<hbool_t[]> data(new hbool_t[values.size()]);
+  for (size_t i = 0; i < values.size(); ++i) {
+    data[i] = values[i];
+  }
+  HDF5WriteAttribute(name, values.size(), data.get(), location);
+}
+
+} // namespace HDF5
+
+namespace {
+
+std::string PackStrings(const std::vector<std::string> &strs, char delimiter) {
+  std::string pack;
+  for (const auto &s : strs) {
+    pack += s + delimiter;
+  }
+  return pack;
+}
+
+std::vector<std::string> UnpackStrings(const std::string &pack, char delimiter) {
+  std::vector<std::string> unpack;
+  const char *curr = pack.data();
+  const char *const end = curr + pack.size();
+
+  while (curr < end) {
+    const auto tab = strchr(curr, delimiter);
+    if (tab == nullptr) {
+      std::stringstream msg;
+      msg << "### ERROR: Pack string does not end with delimiter" << std::endl;
+      PARTHENON_FAIL(msg);
+    }
+
+    if (tab == curr) {
+      unpack.push_back("");
+    } else {
+      unpack.push_back(std::string(curr, tab - curr));
+      curr = tab + 1;
+    }
+  }
+
+  return unpack;
+}
```

Comment:
Wouldn't it be a little simpler to use something like this:

```
std::string input = "abc,def,ghi";
std::istringstream ss(input);
std::vector<std::string> unpack;
std::string token;
while(std::getline(ss, token, ',')) {
    unpack.emplace_back(token);
}
```

You could just check the last character to see if it is a delimiter.

Reply:
That's a good idea, thanks!